From 0a40a57a15433e93ed9eb3c5c292ae83a75d65e6 Mon Sep 17 00:00:00 2001
From: Wyon Bi <bivvy.bi@rock-chips.com>
Date: Mon, 5 Mar 2018 08:39:33 +0800
Subject: [PATCH] phy/rockchip: mipi-dphy: add da_pwrok handling

we can power off the da_pwrok to reduce power consumption.

Change-Id: Ie08af149e74408e57750a186cf16d5adf4b3cfb7
Signed-off-by: Wyon Bi <bivvy.bi@rock-chips.com>
---
 .../phy/rockchip/phy-rockchip-inno-mipi-dphy.c   | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/phy/rockchip/phy-rockchip-inno-mipi-dphy.c b/drivers/phy/rockchip/phy-rockchip-inno-mipi-dphy.c
index 594e0ee26388..18716a10e9f1 100644
--- a/drivers/phy/rockchip/phy-rockchip-inno-mipi-dphy.c
+++ b/drivers/phy/rockchip/phy-rockchip-inno-mipi-dphy.c
@@ -39,6 +39,8 @@
 
 /* Innosilicon MIPI D-PHY registers */
 #define INNO_PHY_LANE_CTRL	0x00000
+#define PWROK_BP		BIT(1)
+#define PWROK			BIT(0)
 #define INNO_PHY_POWER_CTRL	0x00004
 #define ANALOG_RESET_MASK	BIT(2)
 #define ANALOG_RESET		BIT(2)
@@ -395,6 +397,18 @@ static inline void inno_mipi_dphy_reset(struct inno_mipi_dphy *inno)
 			   DIGITAL_RESET_MASK, DIGITAL_NORMAL);
 }
 
+static inline void inno_mipi_dphy_da_pwrok_enable(struct inno_mipi_dphy *inno)
+{
+	regmap_update_bits(inno->regmap, INNO_PHY_LANE_CTRL,
+			   PWROK_BP | PWROK, PWROK);
+}
+
+static inline void inno_mipi_dphy_da_pwrok_disable(struct inno_mipi_dphy *inno)
+{
+	regmap_update_bits(inno->regmap, INNO_PHY_LANE_CTRL,
+			   PWROK_BP | PWROK, PWROK_BP);
+}
+
 static inline void inno_mipi_dphy_lane_enable(struct inno_mipi_dphy *inno)
 {
 	u8 map[] = {0x44, 0x4c, 0x5c, 0x7c};
@@ -641,6 +655,7 @@ static int inno_mipi_dphy_power_on(struct phy *phy)
 	clk_prepare_enable(inno->h2p_clk);
 	clk_prepare_enable(inno->pclk);
 	pm_runtime_get_sync(inno->dev);
+	inno_mipi_dphy_da_pwrok_enable(inno);
 	inno_mipi_dphy_pll_enable(inno);
 	inno_mipi_dphy_lane_enable(inno);
 	inno_mipi_dphy_reset(inno);
@@ -656,6 +671,7 @@ static int inno_mipi_dphy_power_off(struct phy *phy)
 
 	inno_mipi_dphy_lane_disable(inno);
 	inno_mipi_dphy_pll_disable(inno);
+	inno_mipi_dphy_da_pwrok_disable(inno);
 	pm_runtime_put(inno->dev);
 	clk_disable_unprepare(inno->pclk);
 	clk_disable_unprepare(inno->h2p_clk);
-- 
2.35.3

